### PR TITLE
Always confine cursor when left mouse button is pressed

### LINKED ISSUE #1397: Dragging is discontinued upon leaving the window with raw input enabled
When dragging an element (e.g. a slider bar), the dragging is canceled upon leaving the game window while raw input is enabled. It then starts to drag the first element the cursor hits when reentering the game window. This behavior differentiates from disabled raw input as the dragging lasts as long as the mouse button is being hold (independently from the current mouse position).

A video showing this behavior: https://streamable.com/s9gcw

### PR DESCRIPTION
**Pull request by UselessToucan**
Resolves #1397

### CODE REVIEW

**Comment by peppy** [comment]
Rather than all of these changes (most seem unnecessary?) can't we just add `|| lastStates.Any(s => s.IsAnyButtonDown` to the conditional?

**Comment by UselessToucan** [comment]
No, we can't because mouseInWindow becomes false whenever I leave content area of the window.

**Comment by peppy** [comment]
thus the `||`?

**Comment by UselessToucan** [comment]
> thus the ||?

Don't understand where do you suggest to apply this.

File: osu.Framework/Input/Handlers/Mouse/OpenTKRawMouseHandler.cs [hunk 1]
```diff
@@ -59,9 +60,9 @@ public override bool Initialize(GameHost host)
                         if (!host.Window.Visible || host.Window.WindowState == WindowState.Minimized)
                             return;
 
-                        if (mouseInWindow && host.Window.Focused)
```
Comment (peppy) [inline osu.Framework/Input/Handlers/Mouse/OpenTKRawMouseHandler.cs]: i mean change this to
`if ((mouseInWindow || lastStates.Any(s => s.IsAnyButtonDown)) && host.Window.Focused)`
and remove the rest of the changes

File: osu.Framework/Input/Handlers/Mouse/OpenTKRawMouseHandler.cs [hunk 2]
```diff
@@ -174,7 +175,7 @@ private Vector2 getUpdatedPosition(OpenTK.Input.MouseState state, OpenTKMouseSta
                     currentPosition = lastState.Position + new Vector2(state.X - lastState.RawState.X, state.Y - lastState.RawState.Y) * (float)sensitivity.Value;
 
                     // When confining, clamp to the window size.
-                    if (confineMode.Value == ConfineMouseMode.Always || confineMode.Value == ConfineMouseMode.Fullscreen && windowMode.Value == WindowMode.Fullscreen)
+                    if (confineMode.Value == ConfineMouseMode.Always || confineMode.Value == ConfineMouseMode.Fullscreen && windowMode.Value == WindowMode.Fullscreen || lastStates.Any(s => s.HasAnyButtonPressed))
```
Comment (peppy) [inline osu.Framework/Input/Handlers/Mouse/OpenTKRawMouseHandler.cs]: just use `lastState` here

File: osu.Framework/Input/Handlers/Mouse/OpenTKRawMouseHandler.cs [hunk 3]
```diff
@@ -59,9 +60,9 @@ public override bool Initialize(GameHost host)
                         if (!host.Window.Visible || host.Window.WindowState == WindowState.Minimized)
                             return;
 
-                        if (mouseInWindow && host.Window.Focused)
+                        if ((mouseInWindow || lastStates.Any(s => s.HasAnyButtonPressed)) && host.Window.Focused)
```
Comment (peppy) [inline osu.Framework/Input/Handlers/Mouse/OpenTKRawMouseHandler.cs]: probably more efficient to do `lastStates.LastOrDefault()?.HasAnyButtonPressed == true`

File: osu.Framework/Input/Handlers/Mouse/OpenTKRawMouseHandler.cs [hunk 4]
```diff
@@ -174,7 +175,7 @@ private Vector2 getUpdatedPosition(OpenTK.Input.MouseState state, OpenTKMouseSta
                     currentPosition = lastState.Position + new Vector2(state.X - lastState.RawState.X, state.Y - lastState.RawState.Y) * (float)sensitivity.Value;
 
                     // When confining, clamp to the window size.
-                    if (confineMode.Value == ConfineMouseMode.Always || confineMode.Value == ConfineMouseMode.Fullscreen && windowMode.Value == WindowMode.Fullscreen)
+                    if (confineMode.Value == ConfineMouseMode.Always || confineMode.Value == ConfineMouseMode.Fullscreen && windowMode.Value == WindowMode.Fullscreen || lastState.HasAnyButtonPressed)
```
Comment (peppy) [inline osu.Framework/Input/Handlers/Mouse/OpenTKRawMouseHandler.cs]: tbh i'm not even sure this is needed. it feels fine without.

Comment (UselessToucan) [reply osu.Framework/Input/Handlers/Mouse/OpenTKRawMouseHandler.cs]: I guess it is needed when I have ConfineMouseMode = FullScreen and the game in running in Windowed mode.

Comment (peppy) [reply osu.Framework/Input/Handlers/Mouse/OpenTKRawMouseHandler.cs]: It doesn't feel too bad just letting the cursor drift out of the window. We still have some further work to do on confine modes so it's not a huge deal, but I think we may want to just leave this out for now.

**Pull request merged**